src/crystal/system/unix/io_uring.cr

```diff
@@ -0,0 +1,617 @@
+# application can write requests and the completion queue where the kernel
+# will write one response for each request. The operation must be synchronized
+# carefully as the Kernel is using its own thread to interact with these queues
+# without blocking the application.
+#
+# This class is NOT thread-safe.
+class Crystal::System::IoUring
+  # Obtains runtime information about what io_uring features and operations are available.
+  struct Probe
+    property has_io_uring = false
+    property features = Syscall::IoUringFeatures::None
+    property operations = Syscall::IoUringOpAsFlag::None
+
+    # This method should never raise exceptions
+    def initialize
+      params = Syscall::IoUringParams.new
+      fd = Syscall.io_uring_setup(1, pointerof(params))
+      return if fd < 0
+
+      probe = Syscall::IoUringProbe(255).new
+      ret = Syscall.io_uring_register(fd, Syscall::IoUringRegisterOp::REGISTER_PROBE, pointerof(probe).as(Void*), 255)
+      Syscall.close(fd)
+      return if fd < 0
+
+      @has_io_uring = true
+      @features = params.features
+
+      {probe.ops_len, 64}.min.times do |i|
+        operation = probe.ops[i]
+        next unless operation.flags.supported?
+        flag = Syscall::IoUringOpAsFlag.from_value?(1u64 << operation.op)
+        next unless flag
+        @operations |= flag
+      end
+    end
+  end
+
+  @@probe : Probe?
+  @@available : Bool?
+
+  def self.probe
+    @@probe ||= Probe.new
+  end
+
+  # Returns `true` if there are enough features for Crystal's stdlib.
+  def self.available?
+    available = @@available
+    return available unless available.nil?
+
+    probe = self.probe
+
+    # This should return `true` on Linux 5.6+ unless it was built with io_uring disabled
+    @@available = probe.has_io_uring &&
+                  probe.features.nodrop? &&     # NODROP is required because we do not keep track of inflight requests
+                  probe.features.rw_cur_pos? && # Support reading and writing with the "current" file offset
+                  probe.operations.nop? &&
+                  (probe.operations.read? || probe.operations.readv?) &&
+                  (probe.operations.write? || probe.operations.writev?) &&
+                  probe.operations.timeout? &&
+                  probe.operations.accept? &&
+                  probe.operations.connect?
+  end
+
+  # Creates a new io_uring instance with room for at least `sq_entries_hint`. This value is just
+  # a hint. It will be rounded up to the nearest power of two and it will max at 32768.
+  def initialize(sq_entries_hint : UInt32 = 128)
+    @params = Syscall::IoUringParams.new
+
+    # Just clamp to the maximum value if `sq_entries_hint` is too high.
+    @params.flags |= Syscall::IoUringFlags::CLAMP
+
+    # TODO: Consider using SQPOLL when running as root.
+
+    {% if flag? :preview_mt %}
+      # TODO: Use ATTACH_WQ to share kernel resources between threads.
```

Comment:
Hmm, not so certain about what the correct thing to do there is now. There was **major** changes related to this in .12 and .13, and if I understood things correctly now it only make sense to set that flag under SQ_POLL.

Reply:
Probably the safer action is to never set neither SQPOLL nor ATTACH_WQ for now. It can come as an improvement later.